### DISCUSSION TITLE
How to hollow out a mesh via pyvista

### DISCUSSION THREAD

**Discussion by lovecoding2023** [pyvista/pyvista#3121] [Q&A]
Hi,

I am trying to use pyvista to make a function commonly used in 3D printing, hollowing out a solid mesh. For example, I want to make the solid mesh into a mesh with a wall thickness of only 1mm. I have tried many methods, but unfortunately they are unsuccessful. Can give some tips and help, thank you!

**Comment by akaszynski** [comment]
Could you please upload the mesh you're attempting to hollow out?

**Reply by lovecoding2023** [reply]
Thank you very much for such a quick reply, please check the zip file I just uploaded, thank you!
[hollow_file.zip](https://github.com/pyvista/pyvista/files/9237055/hollow_file.zip)

**Reply by akaszynski** [reply]
It's not perfect, but it's indeed been "hollowed out", centered at the original wall location.

```py
import pyvista
pyvista.set_plot_theme('document')
mesh = pyvista.read('hollow_file.stl')

mesh_out = mesh.copy(deep=True)
mesh_out.points += mesh.point_normals*0.5

mesh_in = mesh.copy(deep=True)
mesh_in.points -= mesh.point_normals*0.5
mesh_in.flip_normals()

out = mesh_out + mesh_in
out.plot(smooth_shading=True, split_sharp_edges=True)
```
![tmp](https://user-images.githubusercontent.com/11981631/182241922-4ae73de5-ed56-4c9b-b231-ef5c94e261bf.png)

```
out.slice().plot(cpos='yz', line_width=3)
```
![tmp2](https://user-images.githubusercontent.com/11981631/182241949-3a7b3e3c-6e50-4759-8579-e2fd97d64a92.png)

**Reply by lovecoding2023** [reply]
Thank you so much :)

**Reply by MatthewFlamm** [reply]
Since the answer provided creates a surface both outside and inside the existing outer surface of the part, this method will change the diameter of the through holes in the part.  You may only want to create a new inner surface and combine with the existing outer surface.

**Reply by lovecoding2023** [reply]
Thank you for your kind guidance！ I'm also following your method, and your guidance has helped me understand more about how pyvista works.
But I have encountered a new problem, the inner surface generated in this way will have many overlapping triangles and intersecting triangles, is there an existing algorithm to remove these？

**Reply by lovecoding2023** [reply]
![00](https://user-images.githubusercontent.com/50272068/182426537-80a85be2-4583-460c-8dc1-bf20724a6d8d.png)

**Reply by akaszynski** [reply]
I'd try [pymeshfix](https://pymeshfix.pyvista.org/). It's a bit finicky, but it might work.

**Reply by lovecoding2023** [reply]
Thanks, Can I ask one more question? how can we give the other side of the normal a different color?

**Reply by akaszynski** [reply]
Good question!

I don't think we support that at the moment, but I think you could get it working by modifying the actor that comes off `add_mesh` with:
https://discourse.vtk.org/t/darker-version-of-lookup-table-for-backfaces-of-vtkpolydata/2108/8

There's a bazillion features we need to add; if you do find a way to get this to work we'd love for you to add it.

Also, if you have a fully working solution we'd love to document it.

**Reply by lovecoding2023** [reply]
I seem to have found a solution to avoid generating the overlapping triangles and intersecting triangles. Before this step “mesh_in.points -= mesh.point_normals*0.5”, change all the edge to an oblique angle. Is there an algorithm in pyvista that can transform it into an oblique angle?
![edge](https://user-images.githubusercontent.com/50272068/183285570-3ef07134-c599-4691-832d-3c05db26fd86.png)

**Reply by akaszynski** [reply]
I think what you're looking for (CAD) is best achieved by a library like [cadquery](https://cadquery.readthedocs.io/en/latest/). It's great, really great designed for working with CAD and generating STLs.

**Reply by lovecoding2023** [reply]
thanks:)
